engine/table/src/main/java/io/deephaven/engine/table/impl/locations/local/URIStreamKeyValuePartitionLayout.java

```diff
@@ -0,0 +1,132 @@
+        this.maxPartitioningLevels = Require.geqZero(maxPartitioningLevels, "maxPartitioningLevels");
+        this.fileSeparator = fileSeparator;
+    }
+
+    public String toString() {
+        return URIStreamKeyValuePartitionLayout.class.getSimpleName() + '[' + tableRootDirectory + ']';
+    }
+
+    /**
+     * Find the keys in the given URI stream and notify the observer.
+     */
+    protected final void findKeys(@NotNull final Stream<URI> uriStream,
+            @NotNull final Consumer<TLK> locationKeyObserver) {
+        final KeyValuePartitionLayout.LocationTableBuilder locationTableBuilder = locationTableBuilderFactory.get();
+        final Deque<URI> targetURIs = new ArrayDeque<>();
+        final Set<String> takenNames = new HashSet<>();
+        final List<String> partitionKeys = new ArrayList<>();
+        final boolean[] registered = {false}; // Hack to make the variable final
+        uriStream.forEachOrdered(uri -> {
+            final Collection<String> partitionValues = new ArrayList<>();
+            final String fileRelativePath = uri.getPath().substring(tableRootDirectory.getPath().length());
+            getPartitions(fileRelativePath, partitionKeys, partitionValues, takenNames, registered[0]);
+            if (!registered[0]) {
+                locationTableBuilder.registerPartitionKeys(partitionKeys);
+                registered[0] = true;
+            }
+            locationTableBuilder.acceptLocation(partitionValues);
+            targetURIs.add(uri);
+        });
+        final Table locationTable = locationTableBuilder.build();
+        buildLocationKeys(locationTable, targetURIs, locationKeyObserver, keyFactory);
+    }
+
+    private void getPartitions(@NotNull final String path,
+            @NotNull final List<String> partitionKeys,
+            @NotNull final Collection<String> partitionValues,
+            @NotNull final Set<String> takenNames,
+            final boolean registered) {
+        int partitioningColumnIndex = 0;
+        // Split the path to get the subdirectory names
+        final String[] subDirs = path.split(fileSeparator);
+        for (int i = 0; i < subDirs.length - 1; i++) {
+            final String dirName = subDirs[i];
+            if (dirName.isEmpty()) {
+                // Ignore empty directory names
+                continue;
+            }
+            final String[] components = dirName.split("=", 2);
+            if (components.length != 2) {
+                throw new TableDataException("Unexpected directory name format (not key=value) at "
+                        + new File(tableRootDirectory.getPath(), path));
+            }
+            if (partitioningColumnIndex == maxPartitioningLevels) {
+                throw new TableDataException("Too many partitioning levels at " + path + ", maximum " +
+                        "expected partitioning levels are " + maxPartitioningLevels);
+            }
+            // TODO Is this legalizing adding any value?
```

Comment:
Check with Ryan

Reply:
We might have a bug in that we're not filling `takeNames` with the output of previous calls to `legalizeColumnName`. I (for one) assumed that that parameter was input *and* output.

Otherwise, yes, this adds value.